pkg/group/group_integration_test.go

```diff
@@ -0,0 +1,196 @@
+	require.NoError(t, err, "failed to create admin user and group")
+
+	client := inttest.SetupHTTPServer(t, func(engine *gin.Engine) {
+		handler := group.NewHandler(groupService)
+		authentication := TestAuthenticationMiddleware{}
+		authorization := TestAuthorizationMiddleware{}
+		group.Routes(engine, authentication, authorization, handler)
+	})
+
+	var userId string
+	var user *model.User
+	{
+		user, err = userService.FindOrCreate("user@dhis2.org", "oneoneoneoneoneoneone111")
+		require.NoError(t, err)
+		userId = strconv.FormatUint(uint64(user.ID), 10)
+	}
+
+	var groupName string
+	{
+		group, err := groupService.Create("test-group", "test-hostname.com", true)
+		require.NoError(t, err)
+		groupName = group.Name
+	}
+
+	t.Run("CreateGroup", func(t *testing.T) {
+		t.Parallel()
+
+		t.Run("Deployable", func(t *testing.T) {
+			requestBody := strings.NewReader(`{
+				"name": "deployable-test-group",
+				"hostname": "deployable-test-hostname.com",
+				"deployable": true
+			}`)
+
+			var group model.Group
+			client.PostJSON(t, "/groups", requestBody, &group)
+			require.Equal(t, "deployable-test-group", group.Name)
+			require.Equal(t, "deployable-test-hostname.com", group.Hostname)
+			require.Equal(t, true, group.Deployable)
+		})
+
+		t.Run("NonDeployable", func(t *testing.T) {
+			requestBody := strings.NewReader(`{
+				"name": "non-deployable-test-group",
+				"hostname": "non-deployable-test-hostname.com",
+				"deployable": false
+			}`)
+
+			var group model.Group
+			client.PostJSON(t, "/groups", requestBody, &group)
+			require.Equal(t, "non-deployable-test-group", group.Name)
+			require.Equal(t, "non-deployable-test-hostname.com", group.Hostname)
+			require.Equal(t, false, group.Deployable)
+		})
+	})
+
+	t.Run("AddUserToGroup", func(t *testing.T) {
+		t.Parallel()
+
+		t.Run("AddUserToGroup", func(t *testing.T) {
+			path := fmt.Sprintf("/groups/%s/users/%s", groupName, userId)
+
+			client.Do(t, http.MethodPost, path, nil, http.StatusCreated)
+		})
+
+		t.Run("AddUserToGroupNonExistingGroup", func(t *testing.T) {
+			path := fmt.Sprintf("/groups/%s/users/%s", "non-existing-group", userId)
+
+			response := client.Do(t, http.MethodPost, path, nil, http.StatusNotFound)
+
+			require.Equal(t, "group \"non-existing-group\" doesn't exist", string(response))
+		})
+
+		t.Run("AddNonExistingUserToGroup", func(t *testing.T) {
+			nonExistingUserId := uint(123)
+			_, err := userService.FindById(nonExistingUserId)
+			require.Error(t, err, "user already exists")
+			path := fmt.Sprintf("/groups/%s/users/%s", groupName, fmt.Sprintf("%d", nonExistingUserId))
+
+			response := client.Do(t, http.MethodPost, path, nil, http.StatusNotFound)
+
+			require.Equal(t, "failed to find user with id 123", string(response))
+		})
+	})
+
+	t.Run("RemoveUserFromGroup", func(t *testing.T) {
+		t.Parallel()
+
+		t.Run("RemoveUserFromGroup", func(t *testing.T) {
+			err := groupService.AddUser(groupName, user.ID)
```

Comment:
If I am not mistaken we are adding the same user to the same group `AddUserToGroup` is adding the user to. If that is true, we could run into a race condition as we remove the same user from the group in here.

We could maybe
1. rearrange the tests to write them more like a flow of create user, add user to group, get user, get user with details, remove user.
2. Or we use distinct users/groups so such races cannot occur.

I favor 1. as we are more consistent in testing via HTTP, then sometimes via HTTP and sometimes via a service layer.

Reply:
You are in a way testing `groupService.AddUser(groupName, user.ID)` even if its not your intention.

As said I would

> Rearrange the tests to write them more like a flow of create user, add user to group, get user, get user with details, remove user.

Maybe https://github.com/dhis2-sre/im-manager/blob/b60455c6d6253d393febac1b368931a7a42101d3/pkg/user/user_integration_test.go#L179-L207 helps as an example.

- In a `AsAdmin/Success` test (these tests could be structured as subtests of `TestGroupHandler`). Then in blocks not subtests as these should be executed sequentially
- Create `user1`
- Create group `x`
- Add `user1` to a group `x`
- Get group `x`
- Get group `x` with users and assert `user1` is in it
- Remove `user1` from the group `x`
- Delete group `x`

This simulates a flow that users might actually go through 😄 The flow will create its own user and group. This way there cannot be a race between different tests trying to add/delete the same user to and from a group.